### PR TITLE
Improve error handling when quota limit gets exceeded on API requests.

### PR DESCRIPTION
**Pull request by phgross**
Fixes https://4teamwork.atlassian.net/browse/CA-2247, different from what is described in the JIRA issue, I also corrected the TUS upload, because that is the one that is used in the gever-ui.

Definition of Done: https://4teamwork.atlassian.net/wiki/spaces/CHX4TW/pages/917562/

## Checklist

_Everything has to be done/checked. Checked but not present means the author deemed it unnecessary._

- [x] Changelog entry
- [x] Link to issue (Jira or GitHub) and backlink in issue (Jira)

_Only applicable should be left and checked._

- API change:
  - [x] API Changelog entry

### CODE REVIEW

File: opengever/api/tus.py [hunk 1]
```diff
@@ -1,17 +1,32 @@
...
 from zExceptions import Forbidden
 from zope.component import getMultiAdapter
 from zope.i18n import translate
-from zope.publisher.interfaces import IPublishTraverse
 from zope.interface import implementer
+from zope.publisher.interfaces import IPublishTraverse
+
+
+@implementer(IPublishTraverse)
+class GeverUploadPatch(UploadPatch):
```
Comment (njohner) [inline opengever/api/tus.py:15]: So basically this one is for mails and UploadPatch below is used for documents?

File: opengever/quota/browser/error.py [hunk 2]
```diff
@@ -8,6 +9,9 @@ class ForbiddenByQuotaView(BrowserView):
 
     def __call__(self):
         # this is an error, we must not commit
```
Comment (njohner) [inline opengever/quota/browser/error.py:11]: comment should be moved down or adapted

File: opengever/api/tests/test_tus.py [hunk 3]
```diff
@@ -136,3 +169,13 @@ def test_cannot_replace_proposal_if_not_docx(self, browser):
...
+
+    @browsing
+    def test_raises_if_quota_is_exceeded(self, browser):
+        self.login(self.regular_user, browser)
+
+        api.portal.set_registry_record(interface=IPrivateFolderQuotaSettings,
+                               name='size_hard_limit', value=1)
+
+        self.assert_tus_upload_fails(
+            self.private_dossier, browser, code=507, reason='Insufficient Storage')
```
Comment (njohner) [inline opengever/api/tests/test_tus.py:181]: Can you test that no child was added in `private_dossier`?

File: opengever/api/tests/test_document.py [hunk 4]
```diff
@@ -223,6 +225,19 @@ def test_does_not_allow_mails_as_file(self, browser):
...
             browser.json['message'])
 
+    @browsing
+    def test_raises_if_quota_is_exceeded(self, browser):
+        self.login(self.regular_user, browser)
+
+        api.portal.set_registry_record(interface=IPrivateFolderQuotaSettings,
+                                       name='size_hard_limit', value=1)
+
+        with browser.expect_http_error(code=507, reason='Insufficient Storage'):
```
Comment (njohner) [inline opengever/api/tests/test_document.py:235]: Here too, check that nothing was added to the `private_dossier`

**Review by njohner — APPROVED** [review]
LGTM, just would like to be sure that we are not committing anything, as we are actually catching the exception.

File: opengever/api/tus.py (hunk 1 above)
Comment (phgross) [reply opengever/api/tus.py:15]: Yes actually it does. But GeverUploadPatch applies to all TUS uploads, even if there would be other file types in the future. The document upload TUS inherits from GeverUploadPatch, so it works for Documents as well.

**Comment by phgross** [comment]
@njohner Thx for the review, I added additional `transaction.aboort()` in both exception handlers, to make sure nothing is comitted.

**Pull request merged**